### ISSUE TITLE
Update/Install fail on Ubuntu 10.04.4 with Vim 7.2 and Ruby 1.8.7.

### ISSUE THREAD

**Issue by sangh** [junegunn/vim-plug#162]
I have a relatively old Ubuntu server (`lsb_release -a` says `Ubuntu 10.04.4 LTS`, Vim 7.2 (it is the `vim-nox` package), and `:ruby puts RUBY_VERSION` prints `1.8.7`).  When doing either a `:PlugUpdate` or `:PlugInstall` I get:

    Error detected while processing function <SNR>9_update..<SNR>9_update_impl:
    line   52:
    E16: Invalid range:       silent 4,$d _
    Press ENTER or type command to continue

Setting `let g:plug_threads=1` lets me do a `:PlugInstall`, but weirdly `:PlugUpdate` fails each individual install with:

    error: unknown option `progress'

But this is because the version of git on that machine (1.7.0.4) doesn't have the `--progress` option in clone and fetch.  Actually I don't know why that is there at all since you can't see it, I may submit a pull request just to remove those.

Anyway, I have no idea on the first error, does anyone?

**Comment by junegunn** [comment]
Thanks for the report and PR. I've fixed the code. Let me know if you see any issues.

**Comment by sangh** [comment]
Thanks!

**Comment by sangh** [comment]
Thank you for looking into this!  The progress part seems to work, but hiding the error (with the silent https://github.com/junegunn/vim-plug/commit/f825e4ef6e1939c960175ed09c85e6d0bbe21c6a) just causes vim-plug to say "Finished!" with a successful exit, and without having done anything.  (Of course setting `:let g:plug_threads=1` and using the non-ruby still works great.)  And `:PlugInstall` does the same thing.  Sorry that didn't fix anything.

**Comment by junegunn** [comment]
Ok, let's try to properly fix it. Could you test if the following version works? I changed the code not to use `--progress` option when `v:version < 703`.

https://raw.githubusercontent.com/junegunn/vim-plug/fix-7.2-162/plug.vim

**Comment by junegunn** [comment]
@sangh Okay, here's the real fix. The following version will use `--progress` option only when git is 1.7.1 or above. Could you test if it works for you?

https://raw.githubusercontent.com/junegunn/vim-plug/773b3c/plug.vim

**Comment by sangh** [comment]
@junegunn, I appreciate you looking into this, but I think there has been a misunderstanding.  There are actually two problems here, the first one (which I thought this task was mostly about, and which still doesn't work for me), and the second one about the `--progress` flag.  The soft fix you had the first time works (in response to the pull request) for the progress flag on that box (which I can get to, by bypassing the first problem by setting plug_threads to 1 and using the the non-ruby version.

The first problem (occurring before the progress flag matters), however is still outstanding in that that line with `silent 4,$ _` is failing.  (With `silent!` I don't see the error, but vim-plug just finishes with success _without_ having done anything at all.)

I'm sorry I conflated the two things into one issue (I thought the progress one would be minor and not really mentioned.)  I can open another issue for that, if you want.

**Comment by junegunn** [comment]
Have you tried the updated version I linked above? The line you mentioned is in catch block. Its purpose is to clean up the buffer when the task is interrupted (ctrl-c) so it is not the cause of the problem.

**Comment by sangh** [comment]
Oh, ok, I didn't get that it was only a cleanup section.  Anyway, yes I did try the second thing you linked above and it seems to do exactly the same thing, it prints this:

     Updated. Elapsed time: 0.243845 sec.
    
    
     - Finishing ... Done!

So then, basically, I have no idea.  What else can I try?

**Comment by junegunn** [comment]
Hmm, it looks like the problem isn't as trivial as I would've liked it to be. I'll try to get Ubuntu 10 and look into it when I get some time. Thanks.

**Comment by sangh** [comment]
No worries.  If you want, I *might* be able to get to a shared tmux session on that box or something, but again, with the progress thing I have a work-around so no big deal.

Thanks again.

**Comment by junegunn** [comment]
I set up a lucid VM and looked into the problem. It turned out that the error was caused by the incompleteness of Ruby interface on Vim 7.2.

On Vim 7.3 or above, the following code correctly translates the given Vim dictionary to Ruby Hash:

```vim
:ruby p VIM::evaluate('{"a": 1}')
{"a"=>1}
```

However, on 7.2, it simply doesn't work and raises an error:

```vim
:ruby p VIM::evaluate('{"a": 1}')
E731: using Dictionary as a String
""
```

:weary: 

It looks like String is the *only* type that it understands. So the only option we have is to write a custom function for converting a string representation of a Vim dictionary into the corresponding Ruby Hash. But it's really hacky and we probably shouldn't. I think I should disallow Ruby installer on 7.2 or below.
